File: functions/src/providers/apifootball.ts
// API-FOOTBALL provider implementation.
//
// Docs: https://www.api-football.com/documentation-v3
// Endpoints used:
//   GET /teams?league={id}&season={year}
//   GET /fixtures?league={id}&season={year}
//
// Stage / bracket-slot mapping for the World Cup 2026:
//   API "round" field looks like:
//     "Group Stage - 1", "Round of 32", "Round of 16",
//     "Quarter-finals", "Semi-finals", "3rd Place Final", "Final"
//   We map round + chronological order to our bracket slots
//   (R32-1..R32-16, R16-1..R16-8, QF-1..QF-4, SF-1..SF-2, THIRD, FINAL).
//
// IMPORTANT: actual round labels can shift between API versions. After running
// `syncFixtures` once, eyeball the Admin → Results tab and adjust the mapping
// here if any fixtures land in the wrong slot.

import axios from "axios";
import type { FixtureProvider, NormalizedFixture, NormalizedTeam } from "./types";

const BASE = "https://v3.football.api-sports.io";

interface Options {
  apiKey: string;
  leagueId: number;
  season: number;
}

export function apiFootballProvider(opts: Options): FixtureProvider {
  const client = axios.create({
    baseURL: BASE,
    headers: { "x-apisports-key": opts.apiKey },
    timeout: 15_000,
  });

  return {
    name: "api-football",

    async fetchTeams(): Promise<NormalizedTeam[]> {
      const { data } = await client.get("/teams", {
        params: { league: opts.leagueId, season: opts.season },
      });
      const items: any[] = data?.response ?? [];
      return items.map((it) => ({
        id: String(it.team.id),
        name: it.team.name,
        shortName: it.team.code ?? it.team.name.slice(0, 3).toUpperCase(),
        flag: it.team.logo ?? "",
      }));
    },

    async fetchFixtures(): Promise<NormalizedFixture[]> {
      const { data } = await client.get("/fixtures", {
        params: { league: opts.leagueId, season: opts.season },
      });
      const items: any[] = data?.response ?? [];
      // Sort chronologically so slot indices are deterministic per round.
      items.sort((a, b) => new Date(a.fixture.date).getTime() - new Date(b.fixture.date).getTime());

      const counters: Record<string, number> = {};
      function nextSlot(stage: string): string {
        counters[stage] = (counters[stage] ?? 0) + 1;
        if (stage === "FINAL" || stage === "THIRD") return stage;
        return `${stage}-${counters[stage]}`;
      }

      const out: NormalizedFixture[] = [];
      for (const it of items) {
        const round: string = it.league?.round ?? "";
        const stage = mapStage(round);
        const slot = stage === "GROUP" ? undefined : nextSlot(stage);

        const status = mapStatus(it.fixture?.status?.short ?? "NS");
        const homeId = it.teams?.home?.id ? String(it.teams.home.id) : null;
        const awayId = it.teams?.away?.id ? String(it.teams.away.id) : null;

        out.push({
          externalId: String(it.fixture.id),
          stage,
          group: stage === "GROUP" ? extractGroupLetter(round) : undefined,
          bracketSlot: slot,
          homeTeamId: homeId,
          awayTeamId: awayId,
          kickoffISO: it.fixture.date,
          venue: it.fixture?.venue?.name ?? undefined,
          status,
          homeGoals: it.goals?.home ?? undefined,
          awayGoals: it.goals?.away ?? undefined,
        });
      }
      return out;
    },
  };
}

function mapStage(round: string): NormalizedFixture["stage"] {
  const r = round.toLowerCase();
  if (r.includes("group")) return "GROUP";
  if (r.includes("round of 32")) return "R32";
  if (r.includes("round of 16")) return "R16";
  if (r.includes("quarter")) return "QF";
  if (r.includes("semi")) return "SF";
  if (r.includes("3rd") || r.includes("third")) return "THIRD";
  if (r.includes("final")) return "FINAL";
  return "GROUP";
}

function extractGroupLetter(round: string): string | undefined {
  // "Group Stage - 1" doesn't include the letter; API also returns groups via /standings.
  // Leave undefined here and let the seed script or admin set group letters from the teams response.
  const m = round.match(/Group ([A-L])/i);
  return m?.[1];
}

function mapStatus(s: string): NormalizedFixture["status"] {
  if (["FT", "AET", "PEN"].includes(s)) return "FINISHED";
  if (["1H", "HT", "2H", "ET", "P", "LIVE"].includes(s)) return "LIVE";
  if (["PST", "CANC", "ABD", "AWD", "WO"].includes(s)) return "POSTPONED";
  return "SCHEDULED";
}

File: backend/app/providers/apifootball.py
"""
football-data.org provider (v4) — free tier covers FIFA World Cup.

Replaces the former API-FOOTBALL provider. The function signatures
(fetch_teams, fetch_fixtures) and return shapes are unchanged so the
rest of the codebase doesn't need to change.
"""

from __future__ import annotations

import re

import httpx

BASE_URL = "https://api.football-data.org/v4"
TIMEOUT = 15.0

# football-data.org competition code for FIFA World Cup
COMPETITION = "WC"


async def fetch_teams(
    *, api_key: str, **_kwargs: object,
) -> list[dict]:
    async with httpx.AsyncClient(
        base_url=BASE_URL,
        headers={"X-Auth-Token": api_key},
        timeout=TIMEOUT,
    ) as client:
        resp = await client.get(f"/competitions/{COMPETITION}/teams")
        resp.raise_for_status()
        data = resp.json()

    teams = data.get("teams", [])
    return [
        {
            "id": str(t["id"]),
            "name": t.get("name", ""),
            "shortName": t.get("tla") or t.get("shortName") or t.get("name", "")[:3].upper(),
            "flag": t.get("crest", ""),
            "group": _extract_group_from_area(t),
        }
        for t in teams
    ]


def _extract_group_from_area(team: dict) -> str | None:
    group = team.get("group")
    if group:
        m = re.search(r"Group ([A-L])", group, re.IGNORECASE)
        return m.group(1) if m else None
    return None


async def fetch_fixtures(
    *, api_key: str, **_kwargs: object,
) -> list[dict]:
    async with httpx.AsyncClient(
        base_url=BASE_URL,
        headers={"X-Auth-Token": api_key},
        timeout=TIMEOUT,
    ) as client:
        resp = await client.get(f"/competitions/{COMPETITION}/matches")
        resp.raise_for_status()
        data = resp.json()

    matches = data.get("matches", [])
    matches.sort(key=lambda m: m.get("utcDate", ""))

    counters: dict[str, int] = {}

    def next_slot(stage: str) -> str:
        counters[stage] = counters.get(stage, 0) + 1
        if stage in ("FINAL", "THIRD"):
            return stage
        return f"{stage}-{counters[stage]}"

    out: list[dict] = []
    for m in matches:
        stage_raw = m.get("stage", "")
        group_raw = m.get("group")
        stage = _map_stage(stage_raw)
        slot = None if stage == "GROUP" else next_slot(stage)
        status = _map_status(m.get("status", "SCHEDULED"))

        home = m.get("homeTeam", {})
        away = m.get("awayTeam", {})
        score = m.get("score", {})
        ft = score.get("fullTime", {})

        home_goals = ft.get("home")
        away_goals = ft.get("away")
        if home_goals is None and status == "FINISHED":
            ht = score.get("halfTime", {})
            home_goals = ht.get("home")
            away_goals = ht.get("away")

        out.append(
            {
                "externalId": str(m["id"]),
                "stage": stage,
                "group": _extract_group_letter(group_raw) if stage == "GROUP" else None,
                "bracketSlot": slot,
                "homeTeamId": str(home["id"]) if home.get("id") else None,
                "awayTeamId": str(away["id"]) if away.get("id") else None,
                "kickoffISO": m.get("utcDate", ""),
                "venue": m.get("venue"),
                "status": status,
                "homeGoals": home_goals,
                "awayGoals": away_goals,
            }
        )
    return out


def _map_stage(stage: str) -> str:
    s = stage.upper().replace("_", " ")
    if "GROUP" in s:
        return "GROUP"
    if "ROUND OF 32" in s or "LAST 32" in s or "ROUND 32" in s:
        return "R32"
    if "ROUND OF 16" in s or "LAST 16" in s or "ROUND 16" in s:
        return "R16"
    if "QUARTER" in s:
        return "QF"
    if "SEMI" in s:
        return "SF"
    if "THIRD" in s or "3RD" in s:
        return "THIRD"
    if "FINAL" in s:
        return "FINAL"
    return "GROUP"


def _extract_group_letter(group_name: str | None) -> str | None:
    if not group_name:
        return None
    m = re.search(r"Group ([A-L])", group_name, re.IGNORECASE)
    if m:
        return m.group(1)
    m = re.search(r"GROUP_([A-L])", group_name, re.IGNORECASE)
    if m:
        return m.group(1)
    return None


def _map_status(s: str) -> str:
    s = s.upper()
    if s in ("FINISHED",):
        return "FINISHED"
    if s in ("IN_PLAY", "PAUSED", "LIVE"):
        return "LIVE"
    if s in ("POSTPONED", "CANCELLED", "SUSPENDED"):
        return "POSTPONED"
    return "SCHEDULED"
